spacy_llm/pipeline/llm.py

```diff
@@ -238,26 +236,44 @@ def pipe(self, stream: Iterable[Doc], *, batch_size: int = 128) -> Iterator[Doc]
         """
         error_handler = self.get_error_handler()
         for doc_batch in spacy.util.minibatch(stream, batch_size):
-            is_cached = [doc in self._cache for doc in doc_batch]
-            noncached_doc_batch = [
-                doc for i, doc in enumerate(doc_batch) if not is_cached[i]
-            ]
             try:
-                prompts = self._task.generate_prompts(noncached_doc_batch)
-                responses = self._backend(prompts)
-                modified_docs = iter(self._task.parse_responses(doc_batch, responses))
-                for i, doc in enumerate(doc_batch):
-                    if is_cached[i]:
-                        doc = self._cache[doc]
-                        assert isinstance(doc, Doc)
-                        yield doc
-                    else:
-                        doc = next(modified_docs)
-                        self._cache.add(doc)
-                        yield doc
+                yield from iter(self._process_docs(doc_batch))
             except Exception as e:
                 error_handler(self._name, self, doc_batch, e)
 
+    def _process_docs(self, docs: List[Doc]) -> List[Doc]:
+        docs = list(docs)
+        is_cached = [doc in self._cache for doc in docs]
+
+        noncached_doc_batch = [doc for i, doc in enumerate(docs) if not is_cached[i]]
+        if len(noncached_doc_batch) < len(docs):
+            logger.debug(
+                "Found %d docs in cache. Processing %d docs not found in cache",
+                len(docs) - len(noncached_doc_batch),
+                len(noncached_doc_batch),
+            )
+
+        prompts = list(self._task.generate_prompts(noncached_doc_batch))
+        responses = list(self._backend(prompts))
+        for prompt, response, doc in zip(prompts, responses, noncached_doc_batch):
+            logger.debug("Generated prompt for doc: %s", doc.text)
+            logger.debug(prompt)
+            logger.debug("Backend response for doc: %s", doc.text)
+            logger.debug(response)
```

Comment:
@rmitsch @adrianeboyd

Reply:
I think that's a decent compromise for now.

Reply:
I think it might be confusing when the prompt and response values are separate logging calls from the header/doc info (and might end up out of order or interleaved)? I think it might be useful to combine the prompt / response into one line each?